france_connect/clients.py

```diff
@@ -0,0 +1,222 @@
+
+    def get_access_token(self, code: str) -> Tuple[dict, dict]:
+        """Exchange the authorization code for an access token.
+
+        Parameters
+        ----------
+        code: str
+            The authorization code received after successful authentication.
+
+        Returns
+        -------
+        Tuple[dict, dict]
+            Json containing information token and the verified decoded token.
+        """
+        token_url = urllib.parse.urljoin(self.fc_base_url, self.fc_token_url)
+        data = {
+            "grant_type": "authorization_code",
+            "code": code,
+            "redirect_uri": self.post_login_callback_url,
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+        }
+        response = requests.post(
+            token_url,
+            data=data,
+            headers={"Accept": "application/json"},
+            timeout=self.timeout,
+            verify=self.verify_ssl,
+            allow_redirects=self.allow_redirects,
+        )
+        response.raise_for_status()
+        token = response.json()
+        return token, self.verify_access_token(token.get("id_token"))
+
+    def get_user_info(self, token: str) -> dict:
+        """Retrieve user information using `token`."""
+        url = urllib.parse.urljoin(self.fc_base_url, self.fc_userinfo_url)
+        headers = {"Authorization": f"Bearer {token}", "Accept": "application/json"}
+        response = requests.get(
+            url, headers=headers, timeout=self.timeout, verify=self.verify_ssl, allow_redirects=self.allow_redirects
+        )
+        response.raise_for_status()
+        return self.verify_access_token(response.content.decode("utf-8"))
+
+    def get_logout_url(self, id_token: str, state: str, logout_url_params: dict, is_logout: bool = True) -> str:
+        """Compute the logout URL for FranceConnect.
+
+        Parameters
+        ----------
+        id_token: str
+            The id token to use for the logout.
+        state: str
+            The state to use for the logout.
+        logout_url_params: dict
+            The parameters to use for the post logout url.
+        is_logout: bool, optional
+            If the url is for logout or contact, defaults to True.
+        """
+        post_logout_uri_params = urllib.parse.urlencode(logout_url_params)
+        logout_url = self.post_logout_callback_url if is_logout else self.post_contact_callback_url
+
+        if not logout_url:
+            raise ValueError("You must define `post_logout_redirect_uri` or `post_contact_callback_url` for logout")
+
+        parameters = urllib.parse.urlencode(
+            {
+                "id_token_hint": id_token,
+                "state": state,
+                "post_logout_redirect_uri": f"{logout_url}?{post_logout_uri_params}",
+            }
+        )
+        return f"{urllib.parse.urljoin(self.fc_base_url, self.fc_logout_url)}?{parameters}"
```

Comment:
Instead `logout_url_params` and `is_logout` we should have a `logout_callback_url=None` which default to `self.logout_callback_url`.